Feature/jquery Extend jquery api for #21

mitnuh opened the pull request:
re: #21

About the return types - most of the current api returned a plain ol' go type but the attr() method returns a goja type. Which would you prefer I use?

liclac commented:
Good start!

For goja.Value vs plain types; goja.Value is used to be able to return `goja.Undefined()` where jQuery would do so.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +71,60 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+func optionVal(s *goquery.Selection) string {
+	val, exists := s.Attr("value")
+
+	if exists {
```
imo the code is easier to read with no newline above this, same for eg.

```go
thing, err := Fn()
if err != nil {
  // …
}
```

Makes it immediately obvious which lines are related to each other.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +71,60 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+
+			_, exists := s.sel.Attr("multiple")
+
+			if exists {
```
This can be shortened to `if _, exists := s.sel.Attr("multiple"); exists`

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +71,60 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+				return s.rt.ToValue(optionVal(selected))
+			}
+
+		case "":
```
This is kinda redundant, since it just does the same thing as the default case.

mitnuh replied:
Ah yes, that was pretty dumb. Will fix them all and do some more tomorrow, thank you!

mitnuh commented:
Phew for the tip about AssertFunction() in the issue thread. 

I'm unsure how to handle an invalid fn argument in Each(). Wanted throw a js runtime TypeError but ended up using runtime.Interrupt.

liclac commented:
Try panicking with a `rt.NewGoError(err)` instead, that'll work right even inside a JS->Go->JS callback->Go call stack

mitnuh commented:
That sounds perfect, thanks again. I'm done for today, will be doing more tomorrow.

mitnuh commented:
There's a issue with numeric types returned by the data() function. Need the docs data-attr="101" returns int, and values "101.00" and "1E02" remain as strings but numeric values in JSON strings like data-attr='{"id":101}' are delegated to JSON.parse. 

Parsing the json Go's json unmarshal returns a float64, not an issue inside javascript but the  it makes the test look inconsistent and feels wrong.

mitnuh commented:
sorry for the noise. I wanted to cancel a comment I was writing not close the pull request as I'd realised the answer to something while I was asking.

liclac commented:
Hmm, that does sound a little annoying, but the problem should be purely cosmetic - JS only has a single Number type, so within JS, `1 === 1.0`.

mitnuh commented:
Yeah, it was fine in javascript but made the go tests inconsistent. it was a mistake to return an int from convert - returning float is as valid and more consistent. 

So proud to figure that out while typing but not how to close a comment without closing the PR.

ragnarlonn commented:
@mitnuh Hey Tim. Just chatted with @liclac and we're impressed with your work on this so far, plus we realized the task was probably a little bigger than we initially anticipated, so we're going to increase the bounty on it. Just so you know.

mitnuh commented:
Is there a way to make javascript accessors to wrap go functions? Being able to use accessors for the traversal properties on the dom.Element wrapper would be really useful.

liclac commented:
Sure - just cast them to a JS value and they'll do the right thing.

The one pitfall to look out for is that you can't do `(type, error)` or inject contexts in anything but top-level module functions; those are bootstrapped by common.Proxy(), and the reflection-based wrapping that entrails is far too slow to do in a hot path.

mitnuh commented:
I'm missing something then. I'm trying to convert a struct that's a mix of
properties, methods and hopefully properties that are JS getters to Go
funcs. Best I can do so far is make things like elem.childNodes a function
that needs to be used like 'elem.childNodes()' - using it as
elem.childNodes just gets the a function pointer to native code.



On Thu, May 11, 2017 at 2:31 PM, Emily Ekberg <notifications@github.com>
wrote:

> Sure - just cast them to a JS value and they'll do the right thing.
>
> The one pitfall to look out for is that you can't do (type, error) or
> inject contexts in anything but top-level module functions; those are
> bootstrapped by common.Proxy(), and the reflection-based wrapping that
> entrails is far too slow to do in a hot path.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/loadimpact/k6/pull/197#issuecomment-300789990>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACUzK42y1_C_JESetB4MZQxhwG6VKGcfks5r4w3PgaJpZM4NRJCx>
> .
>

mitnuh commented:
I'll tidy up and send some code for you to look at. Might be a while as I've got some admin that needs doing.

liclac commented:
Oh, sorry, I misread your question - unfortunately you can't make accessors, at least not without a whole JS-side kludge :(

mitnuh commented:
np wasn't clear the first time - will work round it for now.

mitnuh commented:
Hi. Would a JS kludge like this be acceptable? 

https://gist.github.com/mitnuh/b5b4bf16ec7c24d0853f5df23bcc8759

Left a note at the bottom of the gist.

liclac commented:
Not a bad idea, go right ahead! Just make sure you compile the JS blurb into a goja.Program in an init() function, or it will be horribly slow.

If you have a moment, could you write a benchmark comparing your JS blurb vs NewObject() and Set()'ing every attribute individually vs that but requiring parentheses on childNodes(), etc? I don't think performance should be a problem here, but I'd like to be sure.

mitnuh commented:
Will do! I'm about to push to current version then do the benchmarks in a jquery-benchmark branch. Will try out a version with accessors, a version with Set()'ing attributes, and a version using a go object and no js kludges.

mitnuh commented:
Performed just like you said - from horribly slow for the uncompiled version to okay (or at least not terrible) for some variations of the kludge. 

Ditching the JS kludge altogether and just rt.toValue the go struct was the only one that was fast (and the only one that can be run over 10000. 

Big hmm at a resource limit, wild guess is memory related (but not directly linked to free system memory though as it barely changed/plenty free/swap space unchanged - possibly how goja garbage collect/limits high memory use??).

The kludge benchmarks would flake out somewhere between 1000 and 5000 iterations (except the Set() attr version you recommended which sometimes completed 10,000).  The unkludged go wrapped manages 50,000 reliably but loses it at 100,000.

Example results - did a few runs each time with small variations. (Removing Object.freeze was more effective than I expected.)
v3 was the original submission, v4 the same with Object.freeze. 
v 1 & v2 are the compiled version
v5 the Set() attribute version you recommended

// v1 compile proto with object.freeze
BenchmarkBuildSpeed/BenchmarkElement100-4           2000000000           0.01 ns/op
BenchmarkBuildSpeed/BenchmarkElement500-4           2000000000           0.01 ns/op
BenchmarkBuildSpeed/BenchmarkElement1000-4          2000000000           0.10 ns/op
BenchmarkBuildSpeed/BenchmarkElement5000-4          2000000000           0.47 ns/op
BenchmarkBuildSpeed/BenchmarkElement10000-4                1    1859772719 ns/op

// v2 without compile proto without object.freeze
BenchmarkBuildSpeed/BenchmarkElement100-4           2000000000           0.01 ns/op
BenchmarkBuildSpeed/BenchmarkElement500-4           2000000000           0.01 ns/op
BenchmarkBuildSpeed/BenchmarkElement1000-4          2000000000           0.07 ns/op
BenchmarkBuildSpeed/BenchmarkElement5000-4          2000000000           0.37 ns/op
BenchmarkBuildSpeed/BenchmarkElement10000-4                1    1477512700 ns/op

// v3 rt.runstring proto with object.freeze
BenchmarkBuildSpeed/BenchmarkElement100-4           2000000000           0.02 ns/op
BenchmarkBuildSpeed/BenchmarkElement500-4           2000000000           0.03 ns/op
BenchmarkBuildSpeed/BenchmarkElement1000-4          1000000000           0.57 ns/op
BenchmarkBuildSpeed/BenchmarkElement5000-4                 1    2941958126 ns/op
BenchmarkBuildSpeed/BenchmarkElement10000-4                1    5828234282 ns/op

// v4 rt.runstring proto without object.freeze
BenchmarkBuildSpeed/BenchmarkElement100-4           2000000000           0.02 ns/op
BenchmarkBuildSpeed/BenchmarkElement500-4           2000000000           0.02 ns/op
BenchmarkBuildSpeed/BenchmarkElement1000-4          2000000000           0.27 ns/op
BenchmarkBuildSpeed/BenchmarkElement5000-4                 1    2670414828 ns/op
BenchmarkBuildSpeed/BenchmarkElement10000-4                1    5366406879 ns/op

// v5 compiled - without accessor - preset props() or wrapped in functions
BenchmarkBuildSpeed/BenchmarkElement100-4           2000000000           0.00 ns/op
BenchmarkBuildSpeed/BenchmarkElement500-4           2000000000           0.00 ns/op
BenchmarkBuildSpeed/BenchmarkElement1000-4          2000000000           0.05 ns/op
BenchmarkBuildSpeed/BenchmarkElement5000-4          2000000000           0.24 ns/op
BenchmarkBuildSpeed/BenchmarkElement10000-4                1    1009235662 ns/op
BenchmarkElement10000-4 sometimes one time out out 3 at 0.48 ns/op

// v6 de-kludged go struct     
BenchmarkBuildSpeed/BenchmarkElement100-4           2000000000           0.00 ns/op
BenchmarkBuildSpeed/BenchmarkElement500-4           2000000000           0.00 ns/op
BenchmarkBuildSpeed/BenchmarkElement1000-4          2000000000           0.00 ns/op
BenchmarkBuildSpeed/BenchmarkElement5000-4          2000000000           0.02 ns/op
BenchmarkBuildSpeed/BenchmarkElement10000-4         2000000000           0.05 ns/op
BenchmarkBuildSpeed/BenchmarkElement500000-4        2000000000           0.29 ns/op
BenchmarkBuildSpeed/BenchmarkElement1000000-4              1    1036606051 ns/op

(ps that was fun thank you :) )

mitnuh commented:
https://github.com/mitnuh/k6/blob/feature/jquery-benchmark/js/modules/k6/html/speed_test.go

The other versions of the benchmark are the previous half dozen commits.

mitnuh commented:
The current version uses a compiled kludge - the benchmarks were a bit meh so would you prefer to lose the kludge and use a go struct? 

I have a problem with Namespace related properties, they're still todo because the Namespace attribute on html.Node is always empty. I'll investigate why again but if you have any ideas... My backup plan if unsuccessful  is imitate NamespaceURI and isDefaultNamespace by parsing the tag & attr names.

Is there anything else that looks wrong or incomplete?

liclac commented:
Okay so, those benchmarks are a good start, but need a bit of work - have a look at [this](https://github.com/loadimpact/k6/blob/2ae2fe177f6160646b6366c5d78d3f679c09b947/bench_test.go) for inspiration. Instead of running the tests manually like that, use relevant `go test` flags, and I'd like to see memory stats as well.

mitnuh commented:
Which flags are relevant? I was manually running 'go test -test.bench "BuildSpeed" -run ^$ -v'. 

Will add the memory stats and redo the tests so you can compare the variants in one run.

liclac commented:
You probably want `-benchtime` to up the test duration a fair bit.

mitnuh commented:
New benchmarks here: https://gist.github.com/mitnuh/6cde5d7f58f64822c6b4778d44b41d63

Yes, benchtime was most useful. The benchmem results were almost always zero but the odd occasions they were greater than zero it was a test which ran particularly slow.

The original kludge seems to have a performance hit at 50,000 iterations. The JS version without accessors is faster and doesn't. But a similar api to that can be done without a JS kludge using the go struct.

I'm not happy with the abstraction and global vars I used to test the four methods in one benchmark.

liclac commented:
Sorry to get all hung up on this, but I don't understand how these numbers are even possible. Go can't count anything below nanosecond precision, so this sounds like most of the iterations are no-ops...? Even just setting a value on a JS object should be a few dozen nanoseconds *at least*. This sounds like something is very wrong with the benchmarks.

mitnuh commented:
No problem, glad you pointed it. Ah, not sure what I've done wrong, will look into it tonight.

mitnuh commented:
Oh dear, that was dumb. I'd been manually requesting the number of iterations on sub-benchmarks instead of using b.N. The benchmark was asking for tens of thousands of ops when it was doing hundreds so when it averaged the time taken for a small number of ops by the expected number the result was almost zero. 

go test -test.benchtime 10s -test.benchmem -test.bench "BuildSpeed" 
BenchmarkBuildSpeed/Go_Struct-4         	 1000000	     11333 ns/op	    3184 B/op	      56 allocs/op
BenchmarkBuildSpeed/Compiled_AccessorWrapper-4         	  100000	    161407 ns/op	   54503 B/op	     690 allocs/op
BenchmarkBuildSpeed/RunString_AccessorWrapper-4        	   30000	    554752 ns/op	  142248 B/op	    2800 allocs/op
BenchmarkBuildSpeed/Compiled_PresetWrapper-4           	  200000	    108956 ns/op	   37765 B/op	     447 allocs/op

liclac commented:
(Late response, was out on vacation until today.)

I think the JS kludge, as elegant as it is, is just too slow, at least until we can get proper support for properties in goja.

mitnuh commented:
Nice, hope you had a good time.

Makes sens just wish I'd thought to check - will remove kludge and send a PR tonight/tomo.

mitnuh commented:
Removed kludge, things that were attributes need to be accessed with (). The Namespace related issues were my misunderstanding as I was trying to define custom namespaces as if it were an xml parser. 

Thank you for being patient as I made mistakes. I'm here to learn from professionals tbh and will change/add whatever you think best.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+
+	selector := def[0].Export()
+	if len(def) == 1 {
+		switch selector.(type) {
```
Assign in the type switch instead!

```go
switch selector := def[0].Export().(type) {}
```

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+			return Selection{}
+		}
+	} else {
+		filter := def[1].String()
```
Assign in the type switch instead!

```go
switch selector := def[0].Export().(type) {}
```

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+	if len(def) > 0 {
+		return Selection{s.rt, s.sel.Slice(start, def[0])}
+	} else {
+		return Selection{s.rt, s.sel.Slice(start, s.sel.Length())}
```
Just have this the default. It kinda makes sense to have it as an if/else, so I'm torn between the two, but my gut tells me it would be better as a default flow. It's that or a `switch len(def) {}` with only 0/default cases and that feels like it would look downright odd...

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+}
+
+func (s Selection) Get(def ...int) goja.Value {
+	if len(def) == 0 {
```
Make this a switch instead.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+func (s Selection) Get(def ...int) goja.Value {
+	if len(def) == 0 {
+		var items []goja.Value
+
```
I feel like this would be better off without the newlines.

liclac commented on js/modules/k6/html/element.go:
```diff
@@ -0,0 +1,357 @@
+	return e.qsel.Is(selector)
+}
+
+//helper methods
```
These would probably look better at the top of the file or in a `util.go` file, and much easier to find.

liclac replied:
Also, comment formatting:

```go
// Helper methods
```

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+	for i := range s.sel.Nodes {
+		items = append(items, Selection{s.rt, s.sel.Eq(i)})
+	}
+	return
```
As cool as it is that you can actually do this, I'd still go with a good ol' `return items` ^^;;

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+
+func (s Selection) ToArray() (items []Selection) {
+	for i := range s.sel.Nodes {
+		items = append(items, Selection{s.rt, s.sel.Eq(i)})
```
You know the length of `items` before you even start if you look at the selection; save some allocations with `items := make([]Selection, len(s.sel.Nodes))` and `items[i] = ...`!

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+	}
+
+	v := def[0].Export()
+	switch v.(type) {
```
Inline it!

```go
switch v := def[0].Export().(type) {
```

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+		if elem, ok := def[0].Export().(Element); ok {
+			return s.sel.IndexOfNode(elem.node)
+		} else {
+			return -1
```
You can either make this the default, or make the `; ok` above into an assignment followed by an `if !ok {`. Either one is fine.

mitnuh replied:
I can't quite remember why I'm double checking the type here tbh. I think it's the remains of how the kludged version of Element was converted and inlining the type switch made the last trace of it disappear!

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+// end result of the following is two strings.Replacer objects
+// Replacer("-a", "A", "-b", "B"..., "-z", "Z") and Replacer("A", "-a",...)
+//to translate to "data-attr-name" to "attrName" and back
+const (
```
Put all of this at the top of the file instead, constants first, then helpers.

But before doing that: can you refactor this to use `strings` functions instead? That'll handle unicode properly as well.

mitnuh replied:
I've run out of brainpower or tea-power or something so I'll finish this bit tomorrow.

Replacing the hyphens in 'data-attr-name' with space, using a Title() function and removing spaces would get to dataAttrName I think. For going the other is it ok to split the string on upper case chars then joining the array with hyphens?

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -44,17 +48,41 @@ type Selection struct {
-func (s Selection) Add(arg goja.Value) Selection {
-	switch val := arg.Export().(type) {
+func (s Selection) emptySelection() Selection {
+	return s.Eq(s.Size()) //ask for out of bounds item to empty selection
```
Comments should be formatted as:

```go
// Ask for out of bounds item for an empty selection.
```

And preferably put before the line, to give you more space for proper sentences.

(Although none of this is admittedly entirely consistent everywhere...)

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+	}
+}
+
+// end result of the following is two strings.Replacer objects
```
Capitalise "End".

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+	return attrToDataName.Replace(attrName)
+}
+
+// return numeric value when the representation is unchanged by conversion to float and back
```
Capitalise "Return", period at the end.

liclac commented on js/modules/k6/html/element.go:
```diff
@@ -0,0 +1,357 @@
+
+func (e Element) Contains(v goja.Value) bool {
+	if other, ok := v.Export().(Element); ok {
+		// when testing if a node contains itself, jquery + goquery Contains() return true, JS return false
```
Uppercase first character.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+	}
+}
+
+//when 0 args, read all data from attributes beginning with "data-".
```
Whitespace between `//` and comment, uppercase first letter. Same for the next line (which also lacks a period at the end).

liclac commented on js/modules/k6/html/html_test.go:
```diff
@@ -118,4 +137,674 @@ func TestParseHTML(t *testing.T) {
+	t.Run("Get", func(t *testing.T) {
+		t.Run("No args", func(t *testing.T) {
+			v, err := common.RunString(rt, `doc.find("body").children().get()`)
+			// 			fmt.Println(fmt.Sprintf("type is: %T\n", v.Export()))
```
That is a lot of spaces.

mitnuh replied:
93!

liclac commented:
I hope I can be a good mentor then!

mitnuh commented:
Been reading through your review. Looks like lots of good advice, thank you! I have most of tomorrow afternoon to work on this but will leave the PR till Monday so I don't tread on your weekend.

mitnuh commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+			return Selection{s.rt, until("")}
+
+		default:
+			panic(s.rt.NewGoError(errors.New("Invalid argument. The selector must be a string or query object")))
```
Query object isn't, jQuery object is what that was based on - eg https://api.jquery.com/add/#add-selection. Using 'jQuery object' felt wrong because it's really goquery... I'll try some variations.

mitnuh commented:
Sorry about lateness. Not quite finished as I still need to switch to "strings" to handle data attribute names.

mitnuh commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +99,438 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+}
+
+func (s Selection) Filter(v goja.Value) Selection {
+	if gojaFn, isFn := goja.AssertFunction(v); isFn {
```
I couldn't find a variation of gojaFn, isFn := goja.AssertFunction(v) that worked in a case statement, am I missing something? 

Ended up using an inlined type switch and moving the AssertFunction outside the switch - it looks less cramped but I'm not sure it's idiomatic.

mitnuh commented on js/modules/k6/html/html.go:
```diff
@@ -44,17 +47,201 @@ type Selection struct {
+	}
+}
+
+func (s Selection) varargFnCall(arg interface{},
```
Changed from goja.Value to interface{} her to fix a problem I ran into last night (not sure how I missed it :\) at line 87/80 of http.go had been using a string selector to Find(). I could have changed Find to a string only - or - replace the goja.Value with an interface{} and type switch as before, I went with the interface{} and had to use exportIfGojaVal. 

Feel uncertain about this (partly because I didn't spot the problem earlier!)...

liclac replied:
You're already doing a type switch, so why not just add another case for `goja.Value` that calls itself again with the exported value?

mitnuh replied:
Ah, that's neater.

Um, that smiley in my comment was meant to be :\\. But the slash was removed and it was in brackets so now it looks glib and like I was happy but I wasn't.

mitnuh commented:
Ended up using a couple of regular expressions and re.ReplaceAllString, looks cleaner plus unicode! Are regexp's ok here?

liclac commented:
Hm, maybe you could use this? It's already a dep anyways, because the JS bridge uses it to name attributes: https://github.com/serenize/snaker

mitnuh commented:
Ok, just need to replace '_' with '-' and it works. Would you to prefer a modified version of snaker in util.go that need tweaking with Replace?

liclac commented:
Eh, if you want, but it really doesn't matter that much. Single-character string replacements are cheap.

mitnuh commented:
Ok, wasn't sure, will leave it then.

mitnuh commented:
Made some progress, it's going to be a heck of a lot of boilerplate. Is this ok and should I carry on?

liclac commented:
This is actually looking really good! Remember that code generation (`go generate`) exists if you can find a use for it, it's a lifesaver for cutting down on repetitive boilerplate, but only if it's something that can be safely/sanely templated.

liclac commented:
Good example: discordgo's autogenerated type switches for handler functions. https://github.com/bwmarrin/discordgo/blob/0993a94b4e1c3291bed2047f583f34792269355c/events.go

And yes, please rebase, CircleCI checks should all pass.

mitnuh commented:
Won't have a chance to try it out till this evening but that's going to be very useful, thank you.

mitnuh commented:
Made a start on a generator in a different branch.

https://github.com/mitnuh/k6/blob/feature/jquery-codegen/js/modules/k6/html/gen/main.go
https://github.com/mitnuh/k6/blob/feature/jquery-codegen/js/modules/k6/html/elements_gen.go

Must hurry to work, will do more on this tonight!

mitnuh commented:
The output of previous generator was broken and it's config was fussy. Using the ast parser like the discordgo example looks more promising.

mitnuh commented:
Ok, this is pretty close. The only things that's missing is the Video/Media element and all attributes marked obsolete on mozilla.org. I plan on adding the Video/media elements tonight.

The code generation helped hugely, thank you. Not sure about the generated test code though - it feels like more like repeatedly testing whether go is copying strings correctly than testing anything useful.

liclac commented:
Yeah, I prefer to use `t.Run()` and lists/maps of test data over codegen'd tests.

mitnuh commented:
Fixed the tests, improved the generator templates and fixed a few typos, feel quite happy with it now. A few things for you to consider: 

* Relative paths in attributes containing paths/url do not match browser behaviour as I could not find a file name to build relative paths from. Could be tricky as it probably requires changes to other modules.
* Methods like relList and classList currently return an array of strings. Browsers return a object with add/remove/get methods which can also be used as an array. There didn't seem to be a way to get this object + array behaviour in goja, if you'd prefer the object or know how to get the dual behaviour this is pretty easy.
* Only elements which have custom attributes/methods are defined, things like span and h1 etc current using the parent Element struct as they have no extra behaviour. Adding a custom elements for every tag would be trivial but seemed almost pointless as the only thing it would change is the Export value in a few tests, feels more consistent though.
* Ah, Circleci really does not like the repetitive strings in the generated code. I'll change the code generator to use constants later.

liclac commented:
:+1:

If you want, you could totally expose the script filename through `common.State`; it's what it's there for.

mitnuh commented:
This should fix the circleci tests. Nice one, plan to do that tonight.

liclac commented on js/modules/k6/html/elements_gen.go:
```diff
@@ -176,18 +245,18 @@ func (e HrefElement) Download() string {
+	case constunsafe_url:
+		return constunsafe_url
+	default: 
+		return constBlank 
```
Instead of all this, just do:

```go
switch attrVal {
  case "no-referrer", "no-referrer-when-downgrade", ...:
    return attrVal
  default:
    return ""
}
```

mitnuh commented:
Simplified the switch statements. Ran go fmt -s manually after go generate to pass circleci tests, wasn't sure whether to automate gofmt call. 

It was kinda misleading for me to say I wanted a file name to build relative paths from as I really wanted to expose the url given to http.get. Looks like it would involve changing Html in http.go to use response.URL as an extra parameter to ParseHTML then stash the URL in every k6/html/Selection object. Changing the Selection object is the part I'm uncomfortable with, not sure how you feel about it.

liclac commented:
Oh, just have at it tbh. You can't change the parse function, but you can change `req.html()` to just do `sel.URL = …`.

mitnuh commented:
You seem less than convinced, perfectly reasonable and a bit of a relief tbh as my spare time has been limited by work for a few weeks now. The current commit is just a rebase to pull in the changes that will hopefully fix the last few circleci errors. I have most of this weekend free to fix/implement anything and everything that needs to be done, thank you for all your help.

mitnuh commented:
Quick update: The final circleci issue should be fixed with the changes I've made locally but need to rebase to push and I've run out of time before the day job. Done most of the url attribute thing and should be able to submit both tonight.

mitnuh commented:
Apologies for the delay,  had to redo some things and I've been short of time to do them.

The url is set res.html() and there's a test for it in http_test - it felt like the right place. Not sure if exposing the url to javascript was the correct decision, should I change it to an unexposed "url" field and add a SetUrl method or something else?

Adding an url field to the Selection object meant changing how the Selection without changing how the Selection object is constructed in ParseHTML, you said not to touch ParseHTML but couldn't see another way to do it - it doesn't change ParseHTML function signature or it's behaviour, hope that's ok.

mitnuh commented:
Hi. I've changed the url field to URL, as you originally suggested, as it obfuscates the property name in javascript very effectively. The generated const names have been slightly changed for readability. The main change is tweaking how attributes which hold url behave - I hadn't noticed that most of them behave differently when an attribute value is an empty string than when an attribute is undefined and they now match browser behaviour.

I'm really happy with this now, unless there's something you'd like me to do it should be ready for another code review. Thank you.

mitnuh commented:
I need to use sel.URL in res.HTML() in http.go. No computer access atm so will fix tonight.

liclac commented:
I just got back from a vacation, and I'm probably going to hold off on this for a couple of days - I want to go through the whole thing before merging, and the sheer size of it means it's gonna take a few hours.

liclac commented:
Okay, slowly working my way through reviewing this… one major thing I have to say is that I don't like having the code generation stuff in the same package as the code it's generating; I'd much prefer it to be in something like `js/modules/k6/html/gen` or somesuch. *(yo dawg, I heard you like packages…)*

Side note: this PR is now roughly 1/5th the size of the entire k6 codebase 😱

mitnuh commented:
Will do, probabky/hopefully this weekend.

Side note; wtf.

p.s. the way you explain why is properly appreciated. thank you.

On Wed, Sep 6, 2017 at 5:23 PM, Emily Ekberg <notifications@github.com>
wrote:

> Okay, slowly working my way through reviewing this… one major thing I have
> to say is that I don't like having the code generation stuff in the same
> package as the code it's generating; I'd much prefer it to be in something
> like js/modules/k6/html/gen or somesuch. *(yo dawg, I heard you like
> packages…)*
>
> Side note: this PR is now roughly 1/5th the size of the entire k6 codebase
> 😱
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/loadimpact/k6/pull/197#issuecomment-327537944>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACUzK_xM-1MT_RRLLdr87dylPfsjQHz5ks5sfscdgaJpZM4NRJCx>
> .
>

mitnuh commented:
Ok. There a few things in the codegen that are a bit messy and could do with some improvement. I'm considering a few things tbh, one is the way it defines the methods to be generated using the funcDefs array - using more structured data would be neater I think. Another one is the template as it's grown in size and mess, I intend to tidy it up some tonight.

The size of the code is shocking and an awful lot to review. Is there anything I could do to help/lighten the load?

liclac commented:
I agree; I actually have a pending review comment about precisely that. The string-based directives could be replaced with inline struct declarations, easily.

And nah, I'm slowly making my way through it. This is good code, it just has a ridiculous API surface, most of which is just *large*, not so much complex.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +264,191 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+func (s Selection) Each(v goja.Value) Selection {
+	gojaFn, isFn := goja.AssertFunction(v)
+	if !isFn {
+		panic(s.rt.NewGoError(errors.New("Argument to each() must be a function")))
```
Instead of these panic calls, you can use `common.Throw(rt, err)`. That's a helper that makes sure not to double-wrap errors.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +264,191 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+		return s.sel.IsSelection(val.sel)
+	}
+
+	gojaFn, isFn := goja.AssertFunction(v)
```
This should just be a default case in the same switch statement. Looks more consistent.

liclac commented on js/modules/k6/html/html.go:
```diff
@@ -71,3 +264,191 @@ func (s Selection) Attr(name string, def ...goja.Value) goja.Value {
+	fn := func(idx int, sel *goquery.Selection) string {
+		if fnRes, fnErr := gojaFn(v, s.rt.ToValue(idx), s.rt.ToValue(sel)); fnErr == nil {
+			return fnRes.String()
+		} else {
```
No need for an else here.

liclac commented on js/modules/k6/html/util.go:
```diff
@@ -0,0 +1,127 @@
+	"github.com/dop251/goja"
+	"github.com/serenize/snaker"
+
+	gohtml "golang.org/x/net/html"
```
This is a lot of import blocks.

liclac commented on js/modules/k6/html/util.go:
```diff
@@ -0,0 +1,127 @@
+)
+
+func attrToProperty(s string) string {
+	idx := strings.Index(s, "-")
```
You could could shorten this to:

```go
if idx := strings.Index(s, "-"); idx != -1 {
    // …
}
return s
```

liclac commented:
That was a surprisingly nitpicky review for such a huge PR, good job!

mitnuh commented on js/modules/k6/html/elements_gen_test.go:
```diff
@@ -0,0 +1,505 @@
+var textTests = []struct {
+	id       string
+	property string
+	data     string
```
Possibly a misplaced comment?

mitnuh commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -0,0 +1,532 @@
+
+var funcDefs = []string{
+	"Href Download string",
+	"Href ReferrerPolicy enum=,no-referrer,no-referrer-when-downgrade,origin,origin-when-cross-origin,unsafe-url",
```
I ended up going with option 1 using the struct. 

`[]struct {
	Elem, Method, Attr, TemplateType string
	Opts                             []string
}`


There's duplication between the Method and Attr fields of the struct because the method name doesn't always map directly to the name of a DOM attribute. The attribute is usually the method name lowercased but counter-examples are "Encoding" -> "enctype" and "AcceptCharset" ->  "accept-charset". 

If you'd like to get rid of the duplication then the Attr field of the struct can be dropped and the Method name can have an optional override like {"FormElement", "AcceptCharset=accept-charset", "string", nil}

It's pretty easy to do, just not sure if you'd prefer the explicit over brevity in this case.

mitnuh commented:
Thank you! It's my fault for leaving so many nits that needed picking!

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -26,204 +27,203 @@ type ElemInfoCollectorState struct {
-	"Track Default bool",
-	"UList Type string",
+// "Elem" is the struct name for an Element. "Method" is the go method name. "Attr" is the name of the DOM attribute the method will access.
+// "TemplateType" is used the "elemFuncsTemplate" and the returnType function - either string
```
Can we make TemplateType a bunch of constants? I'm not entirely happy with all these repeated strings.

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -232,8 +232,7 @@ func main() {
 	if parseErr != nil {
-		log.Fatalf("warning: internal error: could not parse elements.go: %s", parseErr)
-		return
+		log.Fatalf("error: could not parse elements.go\n", parseErr)
```
No need for \n here if you're using logrus, also use `log.WithError(parseErr).Fatal("...")`, etc. Same for all the other log statements.

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -247,133 +246,110 @@ func main() {
-		}
-		for _, opt := range funcDef.ReturnOpts {
-			enumConsts[opt] = toConst(opt)
+	buf := new(bytes.Buffer)
```
bytes.Buffer is best initialized as:

```go
var buf bytes.Buffer
```

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -1,7 +1,9 @@
+	"go/format"
 	"go/parser"
 	"go/token"
 	"log"
```
Use `log "github.com/Sirupsen/logrus"` instead for consistency.

liclac commented:
This is almost perfect now.

mitnuh commented:
No problem, your reviews have been a great help and consistently made this better, thank you.

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -38,192 +49,192 @@ var funcDefs = []struct {
 	Elem, Method, Attr, TemplateType string
```
Make TemplateType a `type TemplateType string` and define constants with that type instead.

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -38,192 +49,192 @@ var funcDefs = []struct {
 	Elem, Method, Attr, TemplateType string
 	Opts                             []string
```
There's still no explanation of what this is or what values are allowed. Please make an enum or string constant type.

mitnuh replied:
I've added some explanations and a a TemplateArg type. Predefining some values in a var block with some comments helped explain things too. Not sure this is exactly what you had in mind but it is clearer what is required.

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -6,10 +6,21 @@ import (
+)
+
+const (
+	stringTemplate   = "string"
```
These should be named like "typeString", etc. for consistency.

liclac commented on js/modules/k6/html/gen/gen_elements.go:
```diff
@@ -6,10 +6,21 @@ import (
+	urlTemplate      = "url"
+	enumTemplate     = "enum"
+	boolTemplate     = "bool"
+	gojaEnumTemplate = "enum-goja"
```
What does this mean?

mitnuh replied:
Basically it's then enum template except instead of returning a string and using the first item in the enum/list of templateargs as default, it returns a goja.Value and uses goja.Undefined as the default.

I've renamed it to nullableEnumTemplate and added comments to explain it.

liclac commented:
FINALLY! 🎉✨

Sorry that it took so long! The sheer size of the code made it somewhat difficult to review orz

mitnuh commented:
Hurrah! But it might have taken even longer without your advice and the code is more consistent and always improved by your reviews. So sorry about the orz. Thank you!

robingustafsson commented:
@mitnuh Sorry to contact you like this, but I can't find another way to reach you as I don't have an email adress or other contact information.

We've introduced a Contributor License Agreement for k6, and are reaching out to all contributors to ask them for their signature.

You can review and sign the CLA here: https://cla-assistant.io/loadimpact/k6

Hope this message reaches you and that there won't be any hinderances to get the CLA signed 😃

You can reach me at robin@loadimpact.com or via k6 Slack (I'm "robin" in there) if you have any questions. Thank you!
